tailcfg/tailcfg.go

```diff
@@ -1596,3 +1604,18 @@ type SSHAction struct {
 	// value.
 	HoldAndDelegate string `json:"holdAndDelegate,omitempty"`
 }
+
+// ServerMetaResponse is the JSON response to a request to /meta?v=<n>
+// over HTTPS (regular TLS) to the Tailscale control plane server,
+// where <n> is the client's CurrentMapRequestVersion.
+type ServerMetaResponse struct {
+	// PublicNacl specifies the server's original NaCl-based machine key.
+	// It will be zero for sufficiently new clients, based on their
+	// advertised "v" parameter (the CurrentMapRequestVersion).
+	// In that case, only the Noise-based transport may be used.
+	PublicKeyNacl key.MachinePublic `json:"publicKeyNacl"`
+
+	// PublicKeyNoise specifies the server's public key for the
+	// Noise-based control plane protocol.
+	PublicKeyNoise key.MachinePublic `json:"publicKeyNoise"`
```

Comment:
So we're still relying on TLS roots for trusting who the server is. I guess that makes sense seeing as we haven't invented some new mechanism. (On my phone so it's hard to cross reference right now.)

But before we go down this road, I'd like to think through the ramifications of routing ts2021 through DERP. My understanding is we're going to make a noise session then do http2 over that, is that right? If so, then all server metadata, except the key used to establish the connection should be done over the noise channel. That means /meta is too generic a term, because no extra metadata should be outside of noise.

Also, should we introduce some separation here for commands that happen over noise and those that don't?

Reply:
This being over TLS is already discussed in the comment above. This is mostly about converting the existing single-line text HTTP response to be JSON instead, so we can add key rotation date fields later. And it lets us combine the two public keys (for now) into one HTTP round trip. 

There are no plans for any other meta.

We can not name this "meta" if you want and only use the existing `/key` path but return JSON instead of a single line when the `?v=N` parameter is present and >= 27.

Then I guess we'd call this `KeyResponse` instead.

Reply:
Yeah I think naming it something else reduces the risk someone accidentally adds any other metadata to this struct in the future. For newer clients, other discussions should happen over noise. Maybe the comment should point that out too.

This struct is very different than the rest of the structs in tailcfg. Everything else is sent/received over a private channel with the server, this one is out in the open.

Reply:
PTAL?